rpcs/wsFetcherService.go

```diff
@@ -27,50 +31,108 @@ import (
 // WsFetcherService exists for the express purpose or providing a global
 // handler for fetcher gossip message response types
 type WsFetcherService struct {
-	log      logging.Logger
-	mu       deadlock.RWMutex
-	fetchers map[protocol.Tag]*WsFetcher
+	log             logging.Logger
+	mu              deadlock.RWMutex
+	pendingRequests map[string]chan WsGetBlockOut
+}
+
+func makePendingRequestKey(target network.UnicastPeer, round basics.Round, tag protocol.Tag) string {
+	return fmt.Sprintf("<%s>:%d:%s", target.GetAddress(), round, tag)
+
 }
 
 func (fs *WsFetcherService) handleNetworkMsg(msg network.IncomingMessage) (out network.OutgoingMessage) {
 	// route message to appropriate wsFetcher (if registered)
+	switch msg.Tag {
+	case protocol.UniCatchupResTag:
+	case protocol.UniEnsBlockResTag:
+	default:
+		fs.log.Warnf("WsFetcherService: no fetcher registered for tag (%v)", msg.Tag)
+		return
+	}
+
+	uniPeer := msg.Sender.(network.UnicastPeer)
+	var resp WsGetBlockOut
+
+	if msg.Data == nil {
+		fs.log.Warnf("WsFetcherService(%v): request failed: catchup response no bytes sent", uniPeer.GetAddress())
+		out.Action = network.Disconnect
+		return
+	}
+
+	if decodeErr := protocol.Decode(msg.Data, &resp); decodeErr != nil {
+		fs.log.Infof("WsFetcherService(%v): request failed: unable to decode message : %v", uniPeer.GetAddress(), decodeErr)
+		out.Action = network.Disconnect
+		return
+	}
+
+	waitKey := makePendingRequestKey(uniPeer, basics.Round(resp.Round), msg.Tag.Complement())
 	fs.mu.RLock()
-	f := fs.fetchers[msg.Tag]
+	f, hasWaitCh := fs.pendingRequests[waitKey]
 	fs.mu.RUnlock()
-	if f == nil {
-		fs.log.Infof("WsFetcherService: no fetcher registered for tag (%v)", msg.Tag)
+	if !hasWaitCh {
+		if resp.Error == "" || len(resp.BlockBytes) == 0 {
```

Comment:
why do we have this conditional?

Reply:
The difference is the error messages. One contains the error and len(resp.BlockBytes) while the other does not.

Reply:
The better question is why do we print the length and error if either are empty... I'm missing something here. The else only triggers if both are non-empty which seems like the best time to print it, but instead it's the opposite?

Reply:
yes, fixed.